core/shared/src/main/scala/fs2/internal/Resource.scala

```diff
@@ -0,0 +1,165 @@
+package fs2.internal
+
+import java.util.concurrent.atomic.AtomicReference
+
+import cats.effect.Sync
+import fs2.internal.Algebra.Token
+
+/**
+  * Represent resource acquired during stream interpretation
+  *
+  * Resource is acquired by Algebra.Acquire and then released either by Algebra.Release or Algebra.CloseScope.
+  *
+  * The acquisition of the Resource has three steps:
+  *
+  * 1. Resource is created and registered with the scope (Algebra.Acquire)
+  * 2. Resource is acquired
+  * 3. Resource `acquired` is invoked to conform acquisition of the resource.
+  *
+  * The reason for this is that during asynchronous stream evaluation, one scope may cause to close the other scope
+  * (i.e merged stream fails while one of the other streams still acquiring asynchronous resources).
+  * In that case, the resource may be `released` before `acquired` was evaluated, and as such
+  * will make that resource to be released immediately when resource will be acquired..
+  *
+  * The resource may be released by any of the following methods:
+  *
+  * (1) The Algebra.Resource is interpreted. In this case resource finalization will be invoked if
+  *     the resource was successfully acquired (the `acquired` was evaluated and resource was not leased other scope)
+  * (2) The owning Scope was closed by `Algebra.CloseScope`. This essentially evaluates `release` of
+  *     the Resource and acts like (1).
+  * (3) The `acquired` was evaluated after scope was `released` by either (1) or (2). In this case
+  *     resource will invoke the finalization immediately if the resource is not leased.
+  * (4) The `cancelLEase` is invoked for previously leased resource. This will invoke resource
+  *     finalization if resource was already acquired and released and there are no more outstanding leases.
+  *
+  * Resources may be leased to other scopes. in that case, each scope must lease resource with `lease` and
+  * when scope is closed (or when the resource lease is no longer required) release the lease with `cancelLease`.
+  *
+  * Note thate every method which as part of its invocation may potentially call resource finalizer
+  * is returning `F[Either[Throwable, Unit]]` instead just simple F[Unit] to make sure the errors when releasing resource
+  * are properly handled.
+  *
+  *
+  */
+private[internal] sealed trait Resource[F[_]] {
+
+  /** if of the resource **/
```

Comment:
s/if/id/ -- also, end the summary doc with a period to be consistent with other doc in project